feat(site): create price mode selector component, services and adapters

nsilva-spryker opened the pull request:
- add new price mode component
- add new service and adapter
- update effect on cart service, update adapter also
- create new cart guard to check if it's empty on site domain
- update other product services to refresh when price mode change

(Unit tests and storybook will be done in another subtask)

closes:[HRZ-89557](https://spryker.atlassian.net/browse/HRZ-89557)



[HRZ-2804]: https://spryker.atlassian.net/browse/HRZ-2804?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[HRZ-89557]: https://spryker.atlassian.net/browse/HRZ-89557?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nsilva-spryker committed:
feat: add first sketch for price mode component, add to preset header

nsilva-spryker committed:
feat: update tsconfig

nsilva-spryker committed:
feat: update services and adapters for price mode, refactor.

nsilva-spryker committed:
feat: add price mode hydration

nsilva-spryker committed:
feat: update unit test

nsilva-spryker committed:
Merge branch 'development' into feat/hrz-89557-price-mode-component

nsilva-spryker committed:
feat: refactor

nsilva-spryker committed:
feat: update price mode change with guards

nsilva-spryker committed:
Merge branch 'development' into feat/hrz-89557-price-mode-component

nsilva-spryker committed:
feat: update services and adapters

nsilva-spryker committed:
feat: update guard logic and refactor component

nsilva-spryker committed:
Merge branch 'development' into feat/hrz-89557-price-mode-component

nsilva-spryker committed:
feat: small refactor

nsilva-spryker committed:
feat: update unit tests

nsilva-spryker committed:
feat: update unit tests

nsilva-spryker committed:
feat: update unit tests

tolerants commented on libs/domain/site/src/services/site.providers.ts:
```diff
@@ -1,27 +1,34 @@
+// import { CartPriceChangeGuard } from '@spryker-oryx/cart';
```
Leftover?

tolerants commented on libs/template/presets/storefront/experience/header.ts:
```diff
@@ -26,11 +26,14 @@ export const HeaderTemplate: ExperienceComponent = {
             rules: [{ style: 'margin-inline-start: auto' }],
           },
         },
+        {
```
We should register it only for 1.1 or later. Please see https://spryker.slack.com/archives/C029XPSBXC6/p1693293466156589

tolerants commented on libs/domain/site/src/services/price-mode/price-mode-hydration.ts:
```diff
@@ -0,0 +1,10 @@
+import { Observable, skip, take } from 'rxjs';
+import { PriceModeService } from './price-mode.service';
+
+export const priceModeHydration = {
```
So it's not accessible outside the package? Why?

nsilva-spryker replied:
It's only used inside `site` package, don't see any benefits to have it outside, but I could be missing something?

supproduction commented on libs/domain/cart/src/services/default-cart.service.ts:
```diff
@@ -111,6 +112,25 @@ export class DefaultCartService implements CartService {
+            cartId,
+            priceMode: event.data.priceMode,
+          })
+          .subscribe();
```
Why do we subscribe in effect?
```
return this.getCart().pipe(switchMap => this.adapter
          .updateCartPriceMode({
            cartId,
            priceMode: event.data.priceMode,
          })
```

supproduction replied:
It's nor fixed

supproduction commented on libs/domain/site/price-mode-selector/price-mode-selector.component.ts:
```diff
@@ -0,0 +1,98 @@
+    const priceModeInjector = this.injector.inject(PriceModeChangeGuard);
+    priceModeInjector[0]
+      ?.isAllowed()
+      .subscribe((value: boolean) => (isAllowed = value));
```
You assign variable for synchrone code inside async operation. It's fully incorrect approach for rxjs pattern

supproduction commented on libs/domain/site/price-mode-selector/price-mode-selector.component.ts:
```diff
@@ -0,0 +1,98 @@
+        optionNetMode?.removeAttribute('active');
+      } else {
+        optionNetMode?.setAttribute('active', 'true');
+        optionGrossMode?.removeAttribute('active');
```
Why do we need it if we have it in template

nsilva-spryker replied:
I couldn't find a better way to stop propagation when user click one of the options without being authorise to do it(cart not empty). So, this basically reset's previous option to the price mode currently selected.

supproduction commented on libs/domain/site/src/services/price-mode/default-price-mode.service.ts:
```diff
@@ -0,0 +1,30 @@
+
+  constructor(
+    protected queryService = inject(QueryService),
+    protected defaultPriceMode = inject('SCOS_PRICE_MODE')
```
Please, use proper approach for injecting\providing env variables. You can check example for Backend url

nsilva-spryker replied:
It's currently using the same approach for other app environments on site domain
```
 readonly SCOS_BASE_URL?: string;

 readonly STORE?: string;
```
why do we need to change for this case?

supproduction replied:
This one is correct approach
<img width="670" alt="image" src="https://github.com/spryker/oryx/assets/16482929/28238609-1385-4421-97e1-29222db2f541">

supproduction replied:
Because all other cases are incorrected and have to be refactored

supproduction commented on libs/domain/site/src/services/price-mode/price-mode.interceptor.ts:
```diff
@@ -0,0 +1,54 @@
+  ) {}
+
+  intercept(req: Request, handle: HttpHandlerFn): Observable<Response> {
+    return this.injector
+      .inject(PriceModeService)
```
Why we can't inject in from contructor?

supproduction commented on libs/domain/site/src/services/price-mode/price-mode.interceptor.ts:
```diff
@@ -0,0 +1,54 @@
+  ];
+
+  constructor(
+    protected SCOS_BASE_URL = inject('SCOS_BASE_URL'),
```
Same as above

supproduction commented on libs/domain/site/src/services/price-mode/price-mode.interceptor.ts:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    urlObject.searchParams.set(this.parameterName, priceMode);
+    return new Request(urlObject.toString(), req);
```
use req.clone instaed

supproduction commented on libs/domain/cart/src/guards/cart-price.guard.ts:
```diff
@@ -0,0 +1,22 @@
+  isAllowed(): Observable<boolean>;
+}
+
+export class CartPriceModeChangeGuard implements Guard {
+  constructor(protected cartService = inject(CartService)) {}
+
+  isAllowed(): Observable<boolean> {
+    return this.cartService.isEmpty();
+  }
+}
```
Do we really need this guard, why we can't just check `this.cartService.isEmpty();` inside component?

nsilva-spryker replied:
Yes we need this guard from cart domain, to check  in site domain for price mode selection if there is a empty cart.
And we can’t import cart service on site domain because of circular dependency.

supproduction replied:
Why we have `interface Guard ` there then?

supproduction replied:
Move `InjectionTokensContractMap` and Guard into PriceModeChangeGuard and rename Gard => PriceModeChangeGuard interface  then

tolerants commented:
Can you please update PR description, I think `(Unit tests and storybook will be done in another subtask)` is outdated already.

Also can you please update PR title: `add new` - I'd say `new` word is redundant, so what do you think about `create price mode selector component`?

supproduction commented:
Also it's incorrect flow. In dev we merge the full story

supproduction commented:
Please, create story branch from dev and redirect this PR in those branch

nsilva-spryker committed:
Merge branch 'development' into feat/hrz-89557-price-mode-component

nsilva-spryker committed:
feat: small refactor

nsilva-spryker commented:
> Also can you please update PR title: `add new` - I'd say `new` word is redundant, so what do you think about `create price mode selector component`?

`Create` sounds good, I also add 'services and adapters', because in fact those were created also.

nsilva-spryker committed:
feat: update some cr comments

nsilva-spryker committed:
feat: update unit test

nsilva-spryker committed:
feat: small refactor

supproduction commented on apps/storefront/.env.template:
```diff
@@ -1,6 +1,10 @@
 
+# The SCOS_PRICE_MODE env variable can be used to initialise the application with default price mode.
+# When this variable is not provided, the backend will set the default one.
+SCOS_PRICE_MODE=
```
```suggestion
PRICE_MODE=
```

supproduction commented on libs/domain/cart/src/services/adapter/cart.adapter.ts:
```diff
@@ -13,6 +14,7 @@ export interface CartAdapter {
   addEntry: (data: AddCartEntryQualifier) => Observable<Cart>;
   deleteEntry: (data: CartEntryQualifier) => Observable<unknown>;
   updateEntry: (data: UpdateCartEntryQualifier) => Observable<Cart>;
+  updateCartPriceMode: (data: UpdateCartPriceModeQualifier) => Observable<Cart>;
```
Let's create it more generic like `update` cart
```suggestion
  update: (data: UpdateCartQualifier) => Observable<Cart>;
```

supproduction commented on libs/domain/cart/src/services/adapter/default-cart.adapter.ts:
```diff
@@ -73,6 +79,39 @@ export class DefaultCartAdapter implements CartAdapter {
+            type: identity.isAuthenticated
+              ? ApiCartModel.UrlParts.Carts
+              : ApiCartModel.UrlParts.GuestCarts,
+            attributes,
```
```suggestion
            attributes: {...data},
```

supproduction commented on libs/domain/cart/src/services/adapter/default-cart.adapter.ts:
```diff
@@ -115,24 +154,32 @@ export class DefaultCartAdapter implements CartAdapter {
+              attributes: {
+                name: 'My Cart',
+                priceMode: priceMode,
+                currency: currency,
```
```suggestion
                priceMode,
                currency,
```

supproduction commented on libs/domain/site/price-mode-selector/price-mode-selector.component.ts:
```diff
@@ -0,0 +1,95 @@
+  static styles = [sitePriceModeSelectorStyles];
+
+  protected injector = resolve(INJECTOR);
+  protected priceModeGuard = this.injector.inject(PriceModeChangeGuard);
```
```suggestion
  protected priceModeGuard = resolve(PriceModeChangeGuard);
```

supproduction commented on libs/domain/site/price-mode-selector/price-mode-selector.component.ts:
```diff
@@ -0,0 +1,95 @@
+          type: AlertType.Error,
+          content: 'Error',
+          subtext: 'Can’t switch price mode when there are items in the cart',
+        });
```
Can we move this notigication error into interceptor or in service?

supproduction commented on libs/domain/site/price-mode-selector/price-mode-selector.component.ts:
```diff
@@ -0,0 +1,95 @@
+          subtext: 'Can’t switch price mode when there are items in the cart',
+        });
+
+        const optionGrossMode = this.getOptionElement(PriceModes.GrossMode);
+        const optionNetMode = this.getOptionElement(PriceModes.NetMode);
+
+        if (this.$current() === PriceModes.GrossMode) {
+          optionGrossMode?.setAttribute('active', 'true');
+          optionNetMode?.removeAttribute('active');
+        } else {
+          optionNetMode?.setAttribute('active', 'true');
+          optionGrossMode?.removeAttribute('active');
+        }
```
yeah, I checked, so we can't normally programatically set state we do it in popover controller automatically. Is that feature or bug? cc @tobi-or-not-tobi

nsilva-spryker committed:
feat: update some cr comments

nsilva-spryker committed:
feat: update price mode inject env variable, refactor guard

nsilva-spryker committed:
Merge branch 'feature/hrz-2804-price-mode-component' into feat/hrz-89557-price-mode-component

supproduction commented on apps/storefront/src/app.ts:
```diff
@@ -15,6 +17,14 @@ const features = [
   .withFeature(features)
+  .withFeature({
+    providers: [
+      {
+        provide: PriceMode,
+        useFactory: () => injectEnv('PRICE_MODE'),
+      },
```
Let's move it to `site` feature

supproduction commented on libs/domain/cart/src/guards/cart-price.guard.ts:
```diff
@@ -0,0 +1,19 @@
+  }
+}
+
+declare global {
+  interface InjectionTokensContractMap {
+    [PriceModeChangeGuard]: Guard;
+  }
+}
```
I would suggest to move it to the `site` package

supproduction commented on libs/domain/cart/src/services/adapter/cart.adapter.ts:
```diff
@@ -13,6 +14,7 @@ export interface CartAdapter {
   addEntry: (data: AddCartEntryQualifier) => Observable<Cart>;
   deleteEntry: (data: CartEntryQualifier) => Observable<unknown>;
   updateEntry: (data: UpdateCartEntryQualifier) => Observable<Cart>;
+  update: (data: UpdateCartPriceModeQualifier) => Observable<Cart>;
```
UpdateCartPriceModeQualifier => UpdateCartQualifier

supproduction commented on libs/domain/site/price-mode-selector/price-mode-selector.component.ts:
```diff
@@ -0,0 +1,89 @@
+      ?.isAllowed()
+      .pipe(take(1))
+      .subscribe((isAllowed: boolean) => {
+        if (isAllowed) {
+          this.priceModeService.set(locale);
+        } else {
+          this.priceModeService.sendNotificationError();
```
I would like to move the whole if else to the service, is it possible?

supproduction replied:
or lets move obs to the signals to avoid `subscription`

nsilva-spryker committed:
feat: update cr comments

supproduction committed:
fix (#745)

fix component

nsilva-spryker committed:
feat: update cart update subscription

supproduction commented on tsconfig.base.json:
```diff
@@ -246,6 +246,10 @@
         "libs/domain/site/notification-center/index.ts"
       ],
       "@spryker-oryx/site/price": ["libs/domain/site/price/index.ts"],
+      // Release only for 1.1.0 version
```
It's local path, please drop it

supproduction replied:
comment I mean

supproduction committed:
feat: adds cart etag interceptor (#746)

adds etag

closes [HRZ-89614](https://spryker.atlassian.net/browse/HRZ-89614)

[HRZ-89614]:
https://spryker.atlassian.net/browse/HRZ-89614?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nsilva-spryker committed:
Merge branch 'feature/hrz-2804-price-mode-component' into feat/hrz-89557-price-mode-component

nsilva-spryker committed:
Merge branch 'feature/hrz-2804-price-mode-component' into feat/hrz-89557-price-mode-component
